pkg/netpol/eval/internal/k8s/policy_connections.go

```diff
@@ -0,0 +1,151 @@
+// Allowed and Denied connections of current PolicyConnections object (admin-network-policy) are non-overridden.
+// note that:
+// 1. the input connections will include only non-empty allowed conns (since its source is netpols);
+// and any connection that is not allowed by the netpols is denied.
+// 2. pass connections in current PolicyConnections object will be determined by the input PolicyConnections parameter.
+func (pc *PolicyConnections) CollectAllowedConnsFromNetpols(npConns *PolicyConnections) {
+	// subtract the denied conns (which are non-overridden) from input conns
+	npConns.AllowedConns.Subtract(pc.DeniedConns)
+	// PASS conns are determined by npConns
+	// currently, npConns.AllowedConns contains:
+	// 1. traffic that was passed by ANPs (if there are such conns)
+	// 2. traffic that had no match in ANPs
+	// so we can update current allowed conns with them
+	pc.AllowedConns.Union(npConns.AllowedConns)
+	// now pc.AllowedConns contains all allowed conns by the ANPs and NPs
+	// the content of pc.Denied and pc.Pass is not relevant anymore;
+	// all the connections that are not allowed by the ANPs and NPs are denied.
+}
+
+// CollectConnsFromBANP updates current PolicyConnections object (which contains collected conns from ANPs)
+// with connections from a BANP.
+// Allowed and Denied connections of current PolicyConnections object (admin-network-policy) are non-overridden.
+// note that:
+// 1. passConns of the input connections will always be empty. (may contain non-empty allowed/ denied conns)
+// 2. pass connections in current PolicyConnections object will be determined by the input PolicyConnections
+// parameter or system-default value.
+// 3. since both ANP and BANP rules are read as-is; any connection that is not mentioned in any of the admin-policies
+// is allowed by default
+func (pc *PolicyConnections) CollectConnsFromBANP(banpConns *PolicyConnections) {
+	// allowed and denied conns of current pc are non-overridden
+	banpConns.AllowedConns.Subtract(pc.DeniedConns)
+	banpConns.DeniedConns.Subtract(pc.AllowedConns)
+	// currently, banpConns.AllowedConns contains:
+	// 1. traffic that was passed by ANPs (if there are such conns)
+	// 2. or traffic that had no match in ANPs
+	// so we can update current allowed conns with them
+	pc.AllowedConns.Union(banpConns.AllowedConns)
+	// also, banpConns.DeniedConns currently contains:
+	// 1. traffic that was passed by ANPs (if there are such conns)
+	// 2. or traffic that had no match in ANPs
+	// so we can update current denied conns with banpConns.DeniedConns
+	pc.DeniedConns.Union(banpConns.DeniedConns)
+	// now Pass conns were handled automatically; pc.PassConns is not relevant anymore.
+	// Pass Conns which are not captured by BANP, will be handled now with all other conns
+	// all conns that are not determined by the ANP and BANP are allowed by default
+	allowedByDefault := common.MakeConnectionSet(true)
+	allowedByDefault.Subtract(pc.DeniedConns)
```

Comment:
I think it's a bit confusing and not accurate to name this var `allowedByDefault`, because not every connection that is not caught as "denied" by ANP, is considered as "allowed by default", it could also be caught as "allowed" by ANP.

Reply:
renamed to `nonCapturedConns` (not captured by policies)

Reply:
but it may include connections captured by ANPs..

Reply:
true (those conns are already in `pc.AllowedConns`)
this variable purpose is to add the conns which were not captured at all (so they are allowed by default).
do you have another suggestion?

Reply:
i can subtract the pc.AllowedConns from it (but they are union-ed at the end, so the subtract is redundant -  meaningless )

Reply:
few comments/questions:

- as for the name, it should indicate what it contains, which is all connections but (anp/banp)-denied, so a short name could be `allButDenied` with a comment explaining this.
- about the function itself: isn't it sufficient to just compute the following in the function body instead:
```
banpConns.DeniedConns.Subtract(pc.AllowedConns)
pc.DeniedConns.Union(banpConns.DeniedConns)
pc.AllowedConns = common.MakeConnectionSet(true).Subtract(pc.DeniedConns)
```

Reply:
you are right those lines actually does the same, will update it

Reply:
done